task_processing/plugins/mesos/execution_framework.py

```diff
@@ -55,17 +57,28 @@ def __init__(
             checkpoint=True,
             role=self.role
         )
-        self.task_queue = Queue(max_task_queue_size)
-        self.task_update_queue = Queue(max_task_queue_size)
+
+        self.task_queue = make_thread_safe_struct(
```

Comment:
Pyrsistent seems like a good option, and I agree there isn't really a need to add locking on top of queues. 

However, I am concerned about how non-queue data structures like dictionaries/map would prevent race conditions. I suppose they are 'thread-safe' in the sense that since threads gain copies, they do not modify the original and thus do not pollute another thread's work. However, should they try to commit changes on the same thing, you get a conflict. As far as I know, pyrsistent doesn't solve this synchronicity issue itself. I think I'd need to implement locking here.

Alternatively, the thought had crossed my mind that if we instead convert to a coroutine model, explicit async programming would make spotting and controlling race conditions much easier. This way, we could avoid worrying about the nature of the data structures themselves, only about when and how they are used.

Thoughts?